assignment2/binarytree.py

```diff
@@ -0,0 +1,143 @@
+            isLeft (optional): a bool, True if a new node should be a left child
+                               and False if it should be a right child. `None`
+                               by default, that means it doesn't matter which
+                               child would be a new node. If `parent = None`
+                               then `isLeft` should be `None`, otherwise it
+                               raised an error.
+
+        Return:
+            None
+        """
+        if self.root is None:
+            self.root = Node(key)
+            self.size += 1
+            return
+
+        if self.find(key):
+            raise ValueError("key is a duplicate, which not allowed")
+
+        if parent is None:
+            if isLeft is not None:
+                raise ValueError("you should specify a parent if you want" +
+                                 "to insert new node as left/right" +
+                                 "child only")
+            self._insert_internal(key, self.root)
+        else:
+            if not self.find(parent):
+                raise ValueError("there is no node", parent, "in the tree")
+            success = self._insert_internal(key, self.root, parent, isLeft)
+            if not success:
+                raise ValueError("impossible to insert a new node under node",
+                                 parent)
+        self.size += 1
+
+    def _insert_internal(self, key, cur, parent = None, isLeft = None):
+        """Internal iplementation of `insert(...)`.
+
+        Args:
+            key:               an integer, new key that should be inserted;
+            cur:               a Node, current node in the tree that could be a
+                               parent of a new one;
+            parent (optional): see insert(...);
+            isLeft (optional): see insert(...).
+
+        Return:
+            a bool, True if the new key was successfully added to the current
+            node or some of its children, and False otherwise.
+        """
+        if cur is None:
+            return False
+
+        if parent is not None:
+            if cur.key != parent:
+                return (self._insert_internal(key, cur.left, parent, isLeft) or
+                        self._insert_internal(key, cur.right, parent, isLeft))
+            return self._insert_child(key, cur, isLeft)
+
+        if self._insert_child(key, cur):
+            return True
+        return (self._insert_internal(key, cur.right) or
+                self._insert_internal(key, cur.left))
+
+    def _insert_child(self, key, cur, isLeft = None):
+        """Helper function for _insert_internal(...).
+
+        Args: the same as in _insert_internal(...).
+
+        Return:
+            a bool, True if the new key was successfully added to the current
+            node, and False otherwise.
+        """
+        if isLeft is not None:
+            if (isLeft and cur.left is not None or
```

Comment:
Moreover, since cur.left is either None or a node (which evaluates to True in Python) you can write:
```python
  if ((isLeft and not cur.left) or
      (not isLeft and not cur.right)):
```